src/main/java/com/snowflake/kafka/connector/internal/SnowflakeIngestionServiceV1.java

```diff
@@ -45,23 +49,52 @@ public class SnowflakeIngestionServiceV1 implements SnowflakeIngestionService {
       String pipeName,
       PrivateKey privateKey,
       String userAgentSuffix) {
+
+    this(
+        stageName,
+        pipeName,
+        createOrThrow(
+            accountName,
+            userName,
+            host,
+            port,
+            connectionScheme,
+            pipeName,
+            privateKey,
+            userAgentSuffix));
+    LOGGER.info("initialized the pipe connector for pipe {}", pipeName);
+  }
+
+  @VisibleForTesting
+  SnowflakeIngestionServiceV1(
+      String stageName, String pipeName, SimpleIngestManager ingestManager) {
     this.stageName = stageName;
     this.pipeName = pipeName;
+    this.ingestManager = ingestManager;
+  }
+
+  private static SimpleIngestManager createOrThrow(
+      String accountName,
+      String userName,
+      String host,
+      int port,
+      String connectionScheme,
+      String pipeName,
+      PrivateKey privateKey,
+      String userAgentSuffix) {
     try {
-      this.ingestManager =
-          new SimpleIngestManager(
-              accountName,
-              userName,
-              pipeName,
-              privateKey,
-              connectionScheme,
-              host,
-              port,
-              userAgentSuffix);
+      return new SimpleIngestManager(
+          accountName,
+          userName,
+          pipeName,
+          privateKey,
+          connectionScheme,
+          host,
+          port,
+          userAgentSuffix);
     } catch (Exception e) {
-      throw SnowflakeErrors.ERROR_0002.getException(e, this.telemetry);
+      throw SnowflakeErrors.ERROR_0002.getException(e, null);
```

Comment:
why is this set to null?

Reply:
it never had a chance to be initialized to something non null - the object was constructed before this.telemetry is assigned - which is initialized later in the code. 
and i had to pass null explicit here, as i had to make the factory method static (to avoid code duplication)

Reply:
ok, i've refactrored builder, so the telemetry would be accessible in the ctorl